packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresPipeline.hpp

```diff
@@ -276,8 +293,9 @@ class GmresPipeline : public Gmres<SC, MV, OP> {
           Teuchos::Range1D index_prev(0, iter+1);
           const MV Qprev  = * (Q.subView(index_prev));
 
+          auto v_iter = Kokkos::subview(vals, std::pair<int, int>(0, iter+2));
           vec_type W = * (V.getVectorNonConst (iter+1));
-          req = Tpetra::idot (vals, Qprev, W);
+          req = Tpetra::idot (v_iter, Qprev, W);
```

Comment:
Hi @iyamazaki !  It looks like your change has a build error:
```
.../packages/belos/tpetra/solvers/Belos_Tpetra_GmresPipeline.hpp:298:47: error: no matching function for call to ‘idot(Kokkos::View<double*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> >&, const Tpetra::MultiVector<>&, BelosTpetra::Impl::GmresPipeline<double, Tpetra::MultiVector<>, Tpetra::Operator<> >::vec_type&)’
           req = Tpetra::idot (v_iter, Qprev, W);
```
The issue could be that `idot` expects a rank-0 `Kokkos::View` if its second and third arguments have type `Tpetra::Vector`.  That's not what you want here -- you want to compute `Qprev^* W`.  You can fix this by casting `W` to `const MV&`:
```cpp
req = Tpetra::idot(v_iter, Qprev, static_cast<const MV&>(W));
```
GCC 4.8.4 might not be as good at deducing template parameters as newer compilers, so sometimes we have to help it along.

Reply:
Thank you, @mhoemmen for the tip. I have not managed to reproduce the compilation errors, but pushed the change you have suggested.